lib/Sugar.ml

```diff
@@ -326,8 +326,17 @@ module Let_binding = struct
           ( { pexp_desc= Pexp_extension ({txt= "extension.local"; _}, PStr [])
             ; _ }
           , [(Nolabel, sbody)] ) ->
-          let sattrs, _ = check_local_attr sbody.pexp_attributes in
-          let sbody = {sbody with pexp_attributes= sattrs} in
+          let islocal, sbody =
+            (* The desugaring is only valid for some patterns. The pattern
+               part must still be rewritten as the parser duplicated the type
+               annotations and extensions into the pattern and the
+               expression. *)
+            match lb_pat.ppat_desc with
+            | Ppat_var _ | Ppat_constraint ({ppat_desc = Ppat_var _;_}, _) ->
```

Comment:
If I understand correctly, this will cause `let (x : string) = local_ "hi" in ...` to be rewritten to `let local_ (x : string) = "hi" in ...`.  But `local_` only works on a let binding if the pattern is exactly a variable.

Reply:
This indeed cause a syntax error but an equivalent syntax exists:

```ocaml
let local_ x : string = local_ "hi"
```

This rewrite could be allowed by improving the formatting function.

Reply:
The formatting function is not the only problem. The AST of these two lines is different:
```ocaml
let local_ x : string = local_ "hi"
let (x : string) = local_ "hi"
```

I added a commit to distinguish between these two cases (and a third one more complex). The problem is fixed but the code is complex and the rewrite don't happen when it could.
To go further, the parser should be changed to place the constraint in the same place in both cases.

Here are the ASTs:

<details>

Line 1:

```
Ptop_def
  [
    structure_item (_none_[0,0+-1]..[0,0+-1]) ghost
      Pstr_value Nonrec
      [
        <def>
          pattern (_none_[0,0+-1]..[0,0+-1]) ghost
            Ppat_var "x" (_none_[0,0+-1]..[0,0+-1]) ghost
          expression (_none_[0,0+-1]..[0,0+-1]) ghost
            Pexp_apply
            expression (_none_[0,0+-1]..[0,0+-1]) ghost
              Pexp_extension "extension.local"
              []
            [
              <arg>
              Nolabel
                expression (_none_[0,0+-1]..[0,0+-1]) ghost
                  attribute "extension.local"
                    []
                  Pexp_constraint
                  expression (_none_[0,0+-1]..[0,0+-1]) ghost
                    Pexp_constant PConst_string("hi",(_none_[0,0+-1]..[0,0+-1]) ghost,None)
                  core_type (_none_[0,0+-1]..[0,0+-1]) ghost
                    Ptyp_constr "string" (_none_[0,0+-1]..[0,0+-1]) ghost
                    []
            ]
      ]
  ]
```

Line 2:

```
Ptop_def
  [
    structure_item (_none_[0,0+-1]..[0,0+-1]) ghost
      Pstr_value Nonrec
      [
        <def>
          pattern (_none_[0,0+-1]..[0,0+-1]) ghost
            Ppat_constraint
            pattern (_none_[0,0+-1]..[0,0+-1]) ghost
              Ppat_var "x" (_none_[0,0+-1]..[0,0+-1]) ghost
            core_type (_none_[0,0+-1]..[0,0+-1]) ghost
              Ptyp_constr "string" (_none_[0,0+-1]..[0,0+-1]) ghost
              []
          expression (_none_[0,0+-1]..[0,0+-1]) ghost
            Pexp_apply
            expression (_none_[0,0+-1]..[0,0+-1]) ghost
              Pexp_extension "extension.local"
              []
            [
              <arg>
              Nolabel
                expression (_none_[0,0+-1]..[0,0+-1]) ghost
                  Pexp_constant PConst_string("hi",(_none_[0,0+-1]..[0,0+-1]) ghost,None)
            ]
      ]
  ]
```

</details>

Reply:
I am surprised by the first AST here.  The parser in the compiler includes a `Ppat_constraint` in the pattern for this example, and I am having trouble finding where the ocamlformat parsers differ in a way that would account for the absence of this node.

Reply:
I don't know what I did wrong in my previous comment but the ASTs are:

line 1:

```
Ptop_def
  [
    structure_item (./test.ml[1,0+0]..[1,0+35])
      Pstr_value Nonrec
      [
        <def> (./test.ml[1,0+0]..[1,0+35])
          pattern (./test.ml[1,0+11]..[1,0+21]) ghost
            attribute "extension.local" (_none_[0,0+-1]..[0,0+-1]) ghost (_none_[0,0+-1]..[0,0+-1]) ghost
              []
            Ppat_constraint
            pattern (./test.ml[1,0+11]..[1,0+12])
              Ppat_var "x" (./test.ml[1,0+11]..[1,0+12])
            core_type (./test.ml[1,0+15]..[1,0+21]) ghost
              Ptyp_poly
              core_type (./test.ml[1,0+15]..[1,0+21])
                Ptyp_constr "string" (./test.ml[1,0+15]..[1,0+21])
                []
          expression (./test.ml[1,0+4]..[1,0+35]) ghost
            Pexp_apply
            expression (_none_[0,0+-1]..[0,0+-1]) ghost
              Pexp_extension "extension.local" (_none_[0,0+-1]..[0,0+-1]) ghost
              []
            [
              <arg>
              Nolabel
                expression (./test.ml[1,0+4]..[1,0+35])
                  attribute "extension.local" (_none_[0,0+-1]..[0,0+-1]) ghost (_none_[0,0+-1]..[0,0+-1]) 
ghost
                    []
                  Pexp_constraint
                  expression (./test.ml[1,0+24]..[1,0+35]) ghost
                    Pexp_apply
                    expression (_none_[0,0+-1]..[0,0+-1]) ghost
                      Pexp_extension "extension.local" (_none_[0,0+-1]..[0,0+-1]) ghost
                      []
                    [
                      <arg>
                      Nolabel
                        expression (./test.ml[1,0+31]..[1,0+35])
                          Pexp_constant
                          constant (./test.ml[1,0+31]..[1,0+35])
                            PConst_string("hi",(./test.ml[1,0+32]..[1,0+34]),None)
                    ]
                  core_type (./test.ml[1,0+15]..[1,0+21])
                    Ptyp_constr "string" (./test.ml[1,0+15]..[1,0+21])
                    []
            ]
      ]
  ]
```

line 2:

```
Ptop_def
  [
    structure_item (./test.ml[2,36+0]..[2,36+30])
      Pstr_value Nonrec
      [
        <def> (./test.ml[2,36+0]..[2,36+30])
          pattern (./test.ml[2,36+4]..[2,36+16])
            Ppat_constraint
            pattern (./test.ml[2,36+5]..[2,36+6])
              Ppat_var "x" (./test.ml[2,36+5]..[2,36+6])
            core_type (./test.ml[2,36+9]..[2,36+15])
              Ptyp_constr "string" (./test.ml[2,36+9]..[2,36+15])
              []
          expression (./test.ml[2,36+19]..[2,36+30]) ghost
            Pexp_apply
            expression (_none_[0,0+-1]..[0,0+-1]) ghost
              Pexp_extension "extension.local" (_none_[0,0+-1]..[0,0+-1]) ghost
              []
            [
              <arg>
              Nolabel
                expression (./test.ml[2,36+26]..[2,36+30])
                  Pexp_constant
                  constant (./test.ml[2,36+26]..[2,36+30])
                    PConst_string("hi",(./test.ml[2,36+27]..[2,36+29]),None)
            ]
      ]
  ]
```

Both cases have a `Ppat_constraint` indeed. One have the constraint on the expression, while the other doesn't.